### ISSUE TITLE
Request.data empty when multipart/form-data POSTed

### ISSUE THREAD

**Issue by CaffeineFueled** [tomchristie/django-rest-framework#3951]
Django==1.9.2
djangorestframework==3.3.2

This is essentially a repost of #3814.  There is already some research and discussion attached to that issue that I will not rehash, but the author closed out the item once a workaround was found.

Is that workaround the long term solution?  Is the answer to never touch the Django Request before DRF has a chance to do its thing?  This can be difficult to control in projects with third party middleware.

The issue seems to be that, when POSTing data, Django is exhausting the underlying request stream if the WSGIRequest is accessed prior to initialize_request in the DRF View dispatcher.  PUTs seem to be unaffected.

A totally contrived example that demonstrates the issue.  First example, presuming no middleware Request meddling...

```python
class TestView(APIView):

    def post(self, request, *args, **kwargs):
        # request.data, .FILES, and .POST are populated
        # request._request.FILES and .POST are empty
        return Response()

    def put(self, request, *args, **kwargs):
        # request.data, .FILES, and .POST are populated
        # request._request.FILES and .POST are empty
        return Response()
```

Second example, forcibly evaluating the Django request before dispatching.

```python
class TestView(APIView):

    def dispatch(self, request, *args, **kwargs):
        p = request.POST  # Force evaluation of the Django request
        return super().dispatch(request, *args, **kwargs)

    def post(self, request, *args, **kwargs):
        # request.data, .FILES, and .POST are empty
        # request._request.FILES and .POST are populated
        return Response()

    def put(self, request, *args, **kwargs):
        # request.data, .FILES, and .POST are populated
        # request._request.FILES and .POST are empty
        return Response()
```

Cheers!

**Comment by Foslock** [comment]
+1 here. The workaround for the previously closed issue is not a solution for the issue at hand. The truth is that there very well is a use case for accessing the Django POST QueryDict in middleware before DRF can consume it.

**Comment by xordoquy** [comment]
As there is for streaming content too which breaks the use of middleware introspecting the request content.
Moreover, if a middleware do look at the request content, it means it'll bypasses at least partially the content negotiation. At this point, it's probably a good option not to use DRF views and explicitly feed the serializer if required.

**Comment by xordoquy** [comment]
I'd also be interested in learning a bit more in your solution about this issue - in particular how to preserve the compatibility with the current behavior.

**Comment by CaffeineFueled** [comment]
The closest I've come to a "solution" is to add an additional check in the `_parse` method of the DRF `Request`:

```python
...
        try:
            parsed = parser.parse(stream, media_type, self.parser_context)
            # If Django got to the POST/FILES data before we did, just use theirs.
            if (len(parsed.data) + len(parsed.files)) < 1:
                parsed.data = self._request.POST
                parsed.files = self._request.FILES
        except:
            # If we get an exception during parsing, fill in empty data and
...
```

I'm not sure how dangerous it is to use Django's POST data directly like this (rather than making a copy).  They seem to be essentially read only once parsed.

While this retains the current Django WSGIRequest behavior, it leaves it in an inconsistent state insofar as POSTs may or may not have POST/FILE data depending on whether or not the request was touched prior to DRF Request parsing.

**Comment by rsalmaso** [comment]
Just hit this on production today...
As workaround, I've put a check in our base GenericAPIView instead of custom middleware
(for compatibility must keep both override methods)

```python
    def initialize_request(self, request, *args, **kwargs):
       request = super().initialize_request(request, *args, **kwargs)
       if request.path.startswith("/api/3/") and request.method == "POST":
          request.method = request.META.get("HTTP_X_HTTP_METHOD_OVERRIDE", request.POST.get("_method", request.method)).upper()
       return request
```

**Comment by alukach** [comment]
Also running into this problem.  I appears to have been introduced in `3.2` (was not experiencing it while using `3.1.1`).

Anyone know if there is a way to reset the `request` object to its original state after examination or possibly duplicate the object without exhausting the POST data stream?

**Comment by dannybrowne86** [comment]
+1
I have run into a similar issue.  I created a new `nested_route` decorator that acts a lot like `list_route` and `detail_route`.  I usually redirect to another viewset to handle the nested path.  The issue was that the stream was being read too early (by the parent viewset `dispatch`), so I needed to ensure that the `initialize_request` function was only called once for a given request.

Working off of @rsalmaso's comment above, I overrode the `initialize_request` method in the children viewsets with the following.
```
def initialize_request(self, request, *args, **kwargs):
        if not isinstance(request, Request):
            request = super().initialize_request(request, *args, **kwargs)
        return request
```

This works fine, but I think having some sort of attribute for either ignoring the `initialize_request` function within `dispatch` or having a global check to only run `initialize_request` iff `isinstance(request, rest_framework.request.Request) == False`.  I'm happy to prepare the PR with tests if that could be acceptable.

**Comment by jab3z** [comment]
I have the same issue. For now I have reverted to drf 3.2

**Comment by LouWii** [comment]
I've spent almost 3 hours searching about why my data wouldn't be parsed correctly. I first thought the issue was on my Angular App code. And I finally found this issue. Damn, almost banged my head on my desk.

The only work around working for me is getting back to 3.2.5. I'm surprised this issue has no real fix ?

**Comment by xordoquy** [comment]
This shows what [open source really is](http://www.commitstrip.com/en/2014/05/07/the-truth-behind-open-source-apps/).

To get this fixed, the first step is to have someone take time and have a deep analysis of when / why and how the behavior change has been introduced. An initial analysis was performed on #3814 though it requires confirmation that the mentioned change did affect the behavior and then identify how the behavior change has been introduced.

Before that, I won't be able to consider the various workarounds as a solution.

**Comment by CaffeineFueled** [comment]
Here's what I've managed to track down.

The change definitely occurred in commit 566812a as part of removing method and content overriding.  Prior to that change, in [request.py](https://github.com/tomchristie/django-rest-framework/blob/721efa28176507434a2f4b006f2f31b9a33ca2aa/rest_framework/request.py), the [_load_data_and_files](https://github.com/tomchristie/django-rest-framework/blob/721efa28176507434a2f4b006f2f31b9a33ca2aa/rest_framework/request.py#L264) method called [_load_method_and_content_type](https://github.com/tomchristie/django-rest-framework/blob/721efa28176507434a2f4b006f2f31b9a33ca2aa/rest_framework/request.py#L279) which called [_perform_form_overloading](https://github.com/tomchristie/django-rest-framework/blob/721efa28176507434a2f4b006f2f31b9a33ca2aa/rest_framework/request.py#L316).  Within [_perform_form_overloading](https://github.com/tomchristie/django-rest-framework/blob/721efa28176507434a2f4b006f2f31b9a33ca2aa/rest_framework/request.py#L316) was a [bit of code](https://github.com/tomchristie/django-rest-framework/blob/721efa28176507434a2f4b006f2f31b9a33ca2aa/rest_framework/request.py#L336) that seeded the DRF Request._data attribute with Request._request.POST (which is the Django WSGIRequest POST).

```python
        # At this point we're committed to parsing the request as form data.
        self._data = self._request.POST
        self._files = self._request.FILES
        self._full_data = self._data.copy()
        self._full_data.update(self._files)
```

This means that, back in [_load_data_and_files](https://github.com/tomchristie/django-rest-framework/blob/721efa28176507434a2f4b006f2f31b9a33ca2aa/rest_framework/request.py#L264), [_parse](https://github.com/tomchristie/django-rest-framework/blob/721efa28176507434a2f4b006f2f31b9a33ca2aa/rest_framework/request.py#L360) was probably never called for multipart POSTs (at least not in the default configuration) and thus DRF did not attempt to re-read the request stream.

In more recent versions of DRF, [_load_data_and_files](https://github.com/tomchristie/django-rest-framework/blob/0c42c742290400e3ee742e77268681b6911604ed/rest_framework/request.py#L238) no longer calls _load_method_and_content_type as that method no longer exists, so now execution is falling through to the call to [_parse](https://github.com/tomchristie/django-rest-framework/blob/0c42c742290400e3ee742e77268681b6911604ed/rest_framework/request.py#L269).  Herein lies the problem.

For multipart POSTs, [_parse](https://github.com/tomchristie/django-rest-framework/blob/0c42c742290400e3ee742e77268681b6911604ed/rest_framework/request.py#L269) is using a [DRF MultiPartParser](https://github.com/tomchristie/django-rest-framework/blob/0c42c742290400e3ee742e77268681b6911604ed/rest_framework/parsers.py#L90) which is using a [Django MultiPartParser](https://github.com/django/django/blob/dafddb6b8c0eb778072bec1ccd536bafad0eb936/django/http/multipartparser.py#L45) which is attempting to re-read the request stream which has previously been exhausted.

So, another potential solution would be to just allow Django to handle the parsing of multipart forms.  This could be achieved by replacing the [MultiPartParser.parse](https://github.com/tomchristie/django-rest-framework/blob/0c42c742290400e3ee742e77268681b6911604ed/rest_framework/parsers.py#L97) method with something more along the lines of:

```python
    def parse(self, stream, media_type=None, parser_context=None):
        """
        Parses the incoming bytestream as a multipart encoded form,
        and returns a DataAndFiles object.

        `.data` will be a `QueryDict` containing all the form parameters.
        `.files` will be a `QueryDict` containing all the form files.
        """
        parser_context = parser_context or {}
        request = parser_context['request']
        _request = request._request
        return DataAndFiles(_request.POST, _request.FILES)
```

This essentially grabs the WSGIRequest that underlies the DRF Request and allows it to perform the request parsing.  What I don't know is, are there any conditions under which Django would use another parser when DRF is expecting a MultiPartParser?

**Comment by totycro** [comment]
I'm depending on this fix. Are there already plans on when to release it?

**Comment by vakorol** [comment]
I experience the same in 3.3.3. Had to roll back to 3.2.4 on Django 1.8.13. A patch for this issue is highly anticipated.

**Comment by bolgradov-modata** [comment]
I met the same issue today. Doing PUT instead of POST is really the solution for now.

**Comment by tomchristie** [comment]
I'll put it on the list, but no promises if it'll make the 3.4.0 cut or not. We'll see.

**Comment by hvdklauw** [comment]
I did some digging on this too as I ran into it as well.

(This is django 1.8)
The DjangoMultiPartParser loops through the upload_handlers and called handle_raw_input which should return something, but none of them do. 

The MemoryFileUploadHandler just set's activated to True and the TemporaryFileUploadHandler doesn't even implement it.

So maybe the handlers being passed are the wrong one, or rest framework is using the DjangoMultiPartParser wrong.

**Comment by abadabanades** [comment]
+1 Answering @alukach, I have the same issue but only when updating to 3.3.0. With 3.2.0 works fine (I had to downgrade)

**Comment by ggilley** [comment]
Ran in to this issue as well :-( Had to drop back to version 3.2

**Comment by hvdklauw** [comment]
Anything I (we as a company) can do to help with this?
This is preventing us from upgrading to drf >= 3.3 and thus is also blocking our django upgrade.

**Comment by tomchristie** [comment]
Well, it's on the list, and being re-raised by folks, so I'd expect to deal with it prior to 3.5, hopefully towards the end of this month.

> Anything I (we as a company) can do to help with this?

This, this, this...

https://fund.django-rest-framework.org/topics/funding/

(We're just a few % short of what we need for baseline sustainability right now, getting over the mark will mean I can focus exclusively on the project itself)

**Comment by hvdklauw** [comment]
So, we're (Zupr) now funding you to work on DRF 🎉 

This issue has the label Needs information and there is a pull request #4026 with a label needs further review.

Now the pull request by @CaffeineFueled currently does not merge cleanly (but only conflicts in the test file) and was created against 3.3.2 (which was the version that had the issue show up for the first time).

I used the pull request version with our code and ran the test suite (1024 tests, ~84% coverage) and all the test passed.

**Comment by tomchristie** [comment]
Fab 😎 .
Consider this the very next thing on my TODO. I'm at PyCon UK right now, might not get a proper chance to dig in until the weekend, but we'll see.

**Comment by tomchristie** [comment]
Fab 😎 .

There's no great resolution possible with this, but we could consider falling back to Django's parsed data.

Consider this the very next thing on my TODO. I'm at PyCon UK right now, might not get a proper chance to dig in until the weekend, but we'll see.

**Comment by tomchristie** [comment]
Now working on this for a 3.4.7 release.

**Comment by njbuch** [comment]
Not sure about this, as the 3.4.7 is released, and it still does not work to me. (Django 1.9.4) ??

**Comment by tomchristie** [comment]
Can anyone else confirm or deny the fix?

@njbuch Are you able to replicate this in a simple test case or easy to replicate setup?

**Comment by njbuch** [comment]
Its deadsimple:
My view:
```
class PhotoList(APIView):
    parser_classes = (MultiPartParser, FormParser)

    def post(self, request, pk, format=None):
        serializer = PhotoSerializer(data=request.data)
        if serializer.is_valid():
            serializer.save(timestamp_recv=datetime.datetime.utcnow().replace(tzinfo=utc))
            return Response(serializer.data, status=status.HTTP_201_CREATED)
        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)

```
My serializer:
```
class PhotoSerializer(serializers.ModelSerializer):
    class Meta:
        model = TrapPhoto
        fields = ('id', 'trap', 'image'
```

And I am posting this:
```
POST [URL REMOVED] HTTP/1.1
Host: [HOST REMOVED]
Authorization: Basic [REMOVED]
User-Agent: StuffYourEarsWithIcecream/2.48.0
Accept: application/json
Content-Length: 21209
Content-Type: multipart/form-data; boundary=------------------------43beaae86b90c67c

--------------------------43beaae86b90c67c
Content-Disposition: form-data; name="image"; filename="test.gif"
Content-Type: image/gif

GIF89aåØ÷       3  f  ™  Ì  ÿ +  +3 +f +™ +Ì +ÿ U  U3 Uf U™ UÌ Uÿ €  €3 €f €™ €Ì €ÿ ª  ª3 ªf ª™ ªÌ ªÿ Õ  Õ3 
[more binary stuff removed]
--------------------------43beaae86b90c67c--

```

The request.data is simply empty in the view-code, where I want to work on the fields on the body, and the image is always null...

**Comment by tomchristie** [comment]
Reopening for further review.
Further comments from other developers also welcome here.
Surprising given that the test case appears to be resolved.

**Comment by tuky** [comment]
we were stalling our upgrade from django 1.8 to 1.10 because of this and causing failing tests. now we upgraded django and DRF and all tests pass.

**Comment by tuky** [comment]
Something like `curl -X POST -F file0=@....pdf http://localhost:8000/upload_file/ -H "Authorization: Token ..."` is indeed working. the endpoint is a `@detail_route(methods=['post'], parser_classes=(FormParser, MultiPartParser))` of a ViewSet.

**Comment by njbuch** [comment]
I tested with another client, and it seems that changing the client might have an impact. I am testing with Fiddler (which fails) but using the [](https://chrome.google.com/webstore/detail/advanced-rest-client/hgmloofddffdnphfgcellkdfbfbjeloo) with the same data succeeds. Tried my planned client code, which is C# on an embedded device, which also fails. I wondering if the Chrome layer in the ARC actually is formally more correct, and that django-rest-framework maybe is too strict.

**Comment by hvdklauw** [comment]
I notice it is now about files being posted, is the problem still that the request.POST is empty? Or is the request.FILES empty?

**Comment by xordoquy** [comment]
It seems the issue i kind of tricky to track according to what I can read here.
As per @njbuch's comment, it seems client related. Therefore informations about the client may help here to track down this issue.
@njbuch, would it be possible to have like the raw http request (headers and the content with a small file for example) ? That might really help.

**Comment by njbuch** [comment]
Does it make sense to do packet sniffing on this?

**Comment by tomchristie** [comment]
@njbuch Would be one option. We can't really do anything about this unless we have a replicable case. If you can demonstrate how to cause an issue using python's `requests` library to form a request that is handled incorrectly then we could re-open the issue, but right now it sounds like a client problem that needs further investigation.

**Comment by njbuch** [comment]
@tomchristie do you have other examples for test-project, I will try to create a project based that? Just to let you know, same client works in earlier versions of DRF.

**Comment by tomchristie** [comment]
Something that demonstrates the issue against the tutorial project would be tremendous: https://github.com/tomchristie/rest-framework-tutorial

**Comment by alukach** [comment]
I was able to confirm the fix with this repo: https://github.com/alukach/drf-test

Thanks

**Comment by tomchristie** [comment]
Thanks @alukach - will look into that!

**Comment by tomchristie** [comment]
Sorry, read the previous comment incorrectly. Let's try that again...

> I was able to confirm the fix

Wonderful, let's close this off then unless there's any further evidence that this is still an issue against master.

**Comment by hvdklauw** [comment]
Weird that it doesn't throw an error on email, can you add the `UserSerializer` code?

**Comment by xordoquy** [comment]
We will need the simplest test case to work on this as it is usually known to be working.

**Comment by rjpruitt16** [comment]
```
class UserSerializer(serializers.HyperlinkedModelSerializer):
    snapcapsules = SnapCapsuleSerializer(
        many=True,
        read_only=True,
        allow_null=True,
    )

    class Meta:
        model = User
        fields = ('snapcapsules', 'url', 'username', 'email', 'password', )
        write_only_fields = ('password',)

    def create(self, validated_data):
        user = User.objects.create(
            username=validated_data['username'],
            email=validated_data['email'],
        )

        user.set_password(validated_data['password'])
        user.save()

        return user

    def update(self, instance, validated_data):
        capsules_data = validated_data.pop('snapcapsules')

        for capsule in capsules_data:
             SnapCapsule.objects.create(user=instance, **capsule)

        return instance
```

**Comment by rjpruitt16** [comment]
Maybe you could teach me to write the test

**Comment by rpkilby** [comment]
Hi @rjpruitt16. General advice is to reduce this to the minimum amount of code to reproduce the test case. For example, are all of the fields required to replicate the bug? Do you need to override both the `create` and `update` methods? Do you need the nested serializer?

From there, the minimal test case is just the necessary inputs to generate the current, unexpected result (inputs such as test data or an entire request object). Then, compare that against your expected result.

The test doesn't need to be perfect - it just needs to demonstrate the failure.

**Comment by SuperMasterBlasterLaser** [comment]
I have this issue too:

    class ImagesViewSet(ModelViewSet):
        serializer_class = ImageSerializer
        queryset = Image.objects.all()
        permission_classes = (IsAuthenticated,)

        parser_classes = (MultiPartParser, FormParser)

        @action(methods=['post'], detail=False,
            permission_classes=[IsAuthenticated])
        def upload_avatar(self, request):
            print(request.data)
            return Response({"image": "ok"})

After making this request:
![image](https://user-images.githubusercontent.com/8684688/41810042-797fd912-7719-11e8-8b48-693d542b4494.png)

My Query dict returns:

    <QueryDict: {}>

Two years passed since this issue.

**Comment by rpkilby** [comment]
> Two years passed since this issue.

All indications are that this was fixed in version `3.4.7`. If you're still having an issue, the best thing you can do is provide an example test case. You might start by using the example tests provided by @alukach.

https://github.com/alukach/drf-test/commit/77f0e540da42f295197c888b0985c782439d6917

**Comment by rjpruitt16** [comment]
I figure out the problem. I am not sure if my headers were wrong, but in PostMan, If I do not provide any headers than it works perfectly.

You can see my Post on Stackoverflow here
https://stackoverflow.com/questions/50959137/django-rest-framework-unable-to-parse-multipart-form-data

**Comment by ydaniels** [comment]
problem still persist even with latest DRF

**Comment by xordoquy** [comment]
@ydaniels please provide a minimal test case that demonstrate the issue. What I get from the few exchanges here is that it was not an issue with DRF.

**Comment by ydaniels** [comment]
@xordoquy  I am not sure this would do but i made quick one pure drf and django https://github.com/ydaniels/drf-upload-test  . It works perfectly using the ViewSet class but not with ModelViewSet class so the issue should probably be from DRF.

**Comment by rpkilby** [comment]
Hi @ydaniels. The `ImageModelViewSet` is implemented incorrectly. I would recommend reading the below to understand the relationship between `create` and `perform_create`.

https://github.com/encode/django-rest-framework/blob/39bd6cc5cb696d4ba3ab1e7066bda2c4331ff17e/rest_framework/mixins.py#L12-L24

You either need to rename `ImageModelViewSet.perform_create` to `ImageModelViewSet.create` or since the implementation is largely identical to the `CreateModelMixin`, you could delete the method entirely.

**Comment by ydaniels** [comment]
@rpkilby noted Thanks

**Comment by Routhinator** [comment]
@xordoquy I believe I have a case of this not working in DRF 3.11.0 - not sure if this is a regression or I'm missing something, but after a few days of reviewing my code I'm fairly positive I have everything configured correctly, and yet `request.data` is always empty.

I've got a stack overflow open here that has all the code snippets.. https://stackoverflow.com/questions/60741507/drf-serializer-update-method-always-has-empty-validated-data-object-when-using-p 

I'm going to hold off opening a new issue until it's confirmed that this is a new bug. 

I'm using Django 2.2 and DRF 3.11.0 - basic modelviewset and a serializer. This should be working.

The post body from angular is in the issue as well.

**Comment by Routhinator** [comment]
Further testing shows that my case only happens when there are files on the multipart form data. If it only contains fields things are parsed correctly.

**Comment by Routhinator** [comment]
Ok, the boundary= part missing from content-type is what was causing this. So I actually have an angular issue, not a django issue.

**Comment by daxitatrell** [comment]
`class Test(APIView):
    api_view = ['PUT']
    parser_classes = (parsers.MultiPartParser, parsers.FormParser,)

    def put(self, request, format=None):
        print(request.data)
        print(request.FILES)
        print(request.POST)
        return Response(request.data, status=status.HTTP_200_OK)`

Hello any one here to solve my issue? I am using Django==2.2 and restframework=3.10. I am not able to get request.data. I am getting empty Queryset in request.data and request.POST. Help me out this problem. Thanks.

**Comment by ABDURAZZOQ-ABDUSALOMOV** [comment]
abc
